lib/ecies/electrum-ecies.js

```diff
@@ -0,0 +1,150 @@
+var bitcoreECIES = require('./bitcore-ecies')
+var errors = require('./errors')
+
+var AESCBC = require('./aescbc')
+
+// Electrum BIE1 ECIES
+// Difference from Original Bitcore ECIES:
+//  BIE1:
+//      1.Secret S is compressed P(33 bytes), while Bitcore's Secret is Px(32 bytes). See ivkEkM.
+//      2.use AES-128-CBC instead of AES-256-CBC.
+//      3.IV is retrived from first 16 bytes of Hashed Secret Buffer. See iv
+//      4.key Encryption(kE) is retrived from 17-32 bytes of Hashed Secret Buffer, instead of 1-32 bytes. See kE
+//      5.a magic word 'BIE1' is used to identify message type.
+//      6.ephemeral key is introduced, so you can encrypt messages only with public key.
+//      7.HMAC is whole message.
+//  Notice:
+//      Electrum does not support shortTag nor noKey
+//      Do NOT use those 2 options if you are to send a message to Electrum
+//      Encrypted message is NOT recoverable if you use ephemeral key
+//  Security:
+//
+
+// Default algorithm is set to BIE1, however original Bitcore ECIES is still preserved.
+var ECIES = function ECIES (opts, algorithm = 'BIE1') {
+  if (algorithm !== 'BIE1') throw new errors.UnsupportAlgorithm(algorithm)
+  if (!(this instanceof ECIES)) {
+    return new ECIES()
+  }
+  // use ephemeral key if privateKey is not set.
+  this._privateKey = new bsv.PrivateKey()
+  this.opts = opts || {}
+  this.opts.ephemeralKey = true
+}
+
+ECIES.prototype.privateKey = function (privateKey) {
+  $.checkArgument(PrivateKey.isValid(privateKey), 'no private key provided')
+
+  this._privateKey = PrivateKey(privateKey.toString()) || null
+  this.opts.ephemeralKey = false
+
+  return this
+}
+
+ECIES.prototype.publicKey = function (publicKey) {
+  $.checkArgument(PublicKey.isValid(publicKey), 'no public key provided')
+
+  this._publicKey = PublicKey(publicKey.toString()) || null
+
+  return this
+}
+
+var cachedProperty = function (name, getter) {
+  var cachedName = '_' + name
+  Object.defineProperty(ECIES.prototype, name, {
+    configurable: false,
+    enumerable: true,
+    get: function () {
+      var value = this[cachedName]
+      if (!value) {
+        value = this[cachedName] = getter.apply(this)
+      }
+      return value
+    }
+  })
+}
+
+cachedProperty('Rbuf', function () {
+  return this._privateKey.publicKey.toDER(true)
+})
+
+cachedProperty('ivkEkM', function () {
+  var r = this._privateKey.bn
+  var KB = this._publicKey.point
+  var P = KB.mul(r)
+  var S = PublicKey(P)
+  var Sbuf = S.toBuffer()
+  return Hash.sha512(Sbuf)
+})
+
+cachedProperty('iv', function () {
+  return this.ivkEkM.slice(0, 16)
+})
+
+cachedProperty('kE', function () {
+  return this.ivkEkM.slice(16, 32)
+})
+
+cachedProperty('kM', function () {
+  return this.ivkEkM.slice(32, 64)
+})
+
+// Encrypts the message (String or Buffer).
+ECIES.prototype.encrypt = function (message, ivbuf) {
+  if (!Buffer.isBuffer(message)) message = Buffer.from(message)
+  var ciphertext = AESCBC.encrypt(message, this.kE, this.iv)
```

Comment:
here is where you simply want to use the .encryptCipherkey method instead of .encrypt ... that way, the original API can stay the same, and we can still distinguish the two different methods.

Reply:
ivbuf here is unused. will delete.